### PR TITLE
OGM-1462 Support indexes with Neo4j

### PR DESCRIPTION
**Pull request by DavideD**
https://hibernate.atlassian.net/browse/OGM-1462

Follows https://github.com/hibernate/hibernate-ogm/pull/1015

The details are in each commit but, in short, I added some testcase and move the the test for getting and controlling the index during test to the delegate classes.

I also change the test cases to make sure that only the expected indexes exist in the database. To make this work we need to delete the indexes and constraint created by previous tests.

Cheers,
Davide

### CODE REVIEW

**Comment by DavideD** [comment]
Jenkins, retest this please

**Comment by DavideD** [comment]
Jenkins, retest this please

**Comment by DavideD** [comment]
Jenkins, rebuild this, please

**Comment by DavideD** [comment]
Jenkins, retest this please

**Comment by DavideD** [comment]
Jenkins, retest this please

**Comment by DavideD** [comment]
Jenkins, retest this please

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/BaseNeo4jSchemaDefiner.java [hunk 1]
```diff
@@ -181,10 +192,42 @@ public void initializeSchema(SchemaDefinitionContext context) {
...
+					}
+				}
+			}
+		}
+	}
+
+	private void addIndex(Label label, Index index) {
+		if ( index != null ) {
+			if ( index.getName() != null ) {
+				log.cannotSetNameForIndex( index.getName() );
```
Comment (fax4ever) [inline neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/BaseNeo4jSchemaDefiner.java]: Why do we have both cannotSetNameForIndex and warnf?

File: neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+	private static final String PERSON_LABEL = Neo4jIndexTest.class.getSimpleName() + "$" + Person.class.getSimpleName();
+
+	@Rule
+	public ExpectedException thrown = ExpectedException.none();
+
+	@Test
+	public void testSuccessfulIndexCreation() {
+		List<Neo4jIndexSpec> indexes = getIndexes();
+		// Note: I'm not using constainsExactly because it makes it harder to figure out which
```
Comment (gsmet) [inline neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java]: s/constainsExactly/containsExactly/

File: neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+		// Index for ids
+		assertThat( indexes ).contains( indexFor( PERSON_LABEL, true, "id" ) );
+		assertThat( indexes ).contains( indexFor( NodeLabel.SEQUENCE.name(), true, "sequence_name" ) );
+
+		assertThat( indexes ).as( "Unexpected number of indexes created" ).hasSize( 6 );
+	}
+
+	@Test
+	public void testSkippedIndexCreation() {
+		// Unique constraints on multiple properties is not supported in Neo4j at the moment
```
Comment (gsmet) [inline neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java]: s/is/are/

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/BaseNeo4jSchemaDefiner.java (hunk 1 above)
Comment (DavideD) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/BaseNeo4jSchemaDefiner.java]: Definitely an error while rebasing. There should only be the cannotSetNameForIndex.

**Comment by DavideD** [comment]
@fax4ever I might not have the time this week to fix these issues.
Could you please fix these minor things and merge the PR, please?

Thanks

**Comment by DavideD** [comment]
@fax4ever Never mind, I already fixed the PR.

Cheers

**Comment by fax4ever** [comment]
I merged it. Thank you

**Pull request merged**